src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(value, null, tail);
+        if (head == null) {
+            head = tail = node;
+        } else {
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validatePositionIndex(index);
+        if (head == null) {
+            Node<T> node = new Node<>(value, null, null);
+            head = tail = node;
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(value, head, null);
+            head = newNode;
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            Node<T> successor = getNodeByIndex(index);
+            Node<T> predecessor = successor.prev;
+            Node<T> newNode = new Node<>(value, successor, predecessor);
+            successor.prev = newNode;
+            if (predecessor != null) {
+                predecessor.next = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateElementIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateElementIndex(index);
+        Node<T> targetNode = getNodeByIndex(index);
+        T oldValue = targetNode.value;
+        targetNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateElementIndex(index);
+        return unlink(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
```

Comment:
remove if-else - adjust internal if's condition instead, it should be null-safe check like one in Objects.equals(look it up - you can't use it directly in this hw yet)